arduino/WandCore/WandCore.ino

```diff
@@ -175,50 +190,135 @@ void setup()
-
   // Use the built-in LED as visual feedback
   pinMode(LED_BUILTIN, OUTPUT);
 
-  // Double flash to indicate end of setup
+  // Flash onboard and RGB LEDs to indicate end of setup
   digitalWrite(LED_BUILTIN, HIGH);
+  setRGBLED(255, 0, 0);
   delay(200);
   digitalWrite(LED_BUILTIN, LOW);
+  setRGBLED(0, 0, 0);
   delay(200);
   digitalWrite(LED_BUILTIN, HIGH);
+  setRGBLED(0, 0, 255);
+  delay(200);
+  digitalWrite(LED_BUILTIN, LOW);
+  setRGBLED(0, 0, 0);
+  delay(200);
+  digitalWrite(LED_BUILTIN, HIGH);
+  setRGBLED(0, 255, 0);
   delay(200);
   digitalWrite(LED_BUILTIN, LOW);
+  setRGBLED(255, 255, 255); // white for power mode, which is default
 
   // testDevices();
 }
 
-void testDevices()
+//void testDevices()
+//{
+//  // Test RGB LED
+//  setRGBLED(255, 0, 0);
+//  delay(1000);
+//  setRGBLED(0, 255, 0);
+//  delay(1000);
+//  setRGBLED(0, 0, 255);
+//  delay(1000);
+//  setRGBLED(0, 0, 0);
+//  delay(1000);
+//
+//  // Test color sensor
+//  setColorSensorLED(true);
+//  delay(1000);
+//  setColorSensorLED(false);
+//  for (int i = 0; i < 5; i++)
+//  {
+//    getRGB(&sensorR, &sensorG, &sensorB);
+//    delay(200);
+//  }
+//}
+
+void loop() 
 {
-  // Test RGB LED
-  setRGBLED(255, 0, 0);
-  delay(1000);
-  setRGBLED(0, 255, 0);
-  delay(1000);
-  setRGBLED(0, 0, 255);
-  delay(1000);
-  setRGBLED(0, 0, 0);
-  delay(1000);
+  ArduinoOTA.handle();
 
-  // Test color sensor
-  setColorSensorLED(true);
-  delay(1000);
-  setColorSensorLED(false);
-  for (int i = 0; i < 5; i++)
+  // If in Color Mode, increment the RGB LED in the loop
+  if (wandMode == 2)
   {
-    getRGB(&sensorR, &sensorG, &sensorB);
-    delay(200);
+    switch (rainbowSection) // implementing the 6 sections here https://academe.co.uk/wp-content/uploads/2012/04/451px-HSV-RGB-comparison.svg_.png
+    {
+      case 1: // red stays up, green goes up, blue stays down
+        setRGBLED(255, rainbowVar, 0);
+        rainbowVar++;
+        if (rainbowVar == 255) rainbowSection++;
+        break;
+      case 2: // red goes down, green stays up, blue stays down
+        setRGBLED(rainbowVar, 255, 0);
+        rainbowVar--;
+        if (rainbowVar == 0) rainbowSection++;
+        break;
+      case 3: // red stays down, green stays up, blue goes up
+        setRGBLED(0, 255, rainbowVar);
+        rainbowVar++;
+        if (rainbowVar == 255) rainbowSection++;
+        break;
+      case 4: // red stays down, green goes down, blue stays up
+        setRGBLED(0, rainbowVar, 255);
+        rainbowVar--;
+        if (rainbowVar == 0) rainbowSection++;
+        break;
+      case 5: // red goes up, green stays down, blue stays up
+        setRGBLED(rainbowVar, 0, 255);
+        rainbowVar++;
+        if (rainbowVar == 255) rainbowSection++;
+        break;
+      case 6: // red stays up, green stays down, blue goes down
+        setRGBLED(255, 0, rainbowVar);
+        rainbowVar--;
+        if (rainbowVar == 0) rainbowSection = 1;
+        break;
+    }

```

Comment:
This does work, but it's probably way easier to convert RBG to HSL and back and just cycle Hue through the hue spectrum (which with a constant Saturation and Lightness is a smooth color loop).

Reply:
I also think this should be abstracted out into the `LEDClient`.
You could also set the speed, etc. This color loop implementation isn't necessarily going to have consistent speed, depending if iterations of `loop()` take different times.

Reply:
abstracting out now, but not updating the algorithm at this time, will mark as TODO